format/nes/nes.go

```diff
@@ -0,0 +1,339 @@
+
+		// byte 8
+		d.FieldU8("prg_ram_size", ramSizer(8))
+
+		// byte 9 - 15
+		d.FieldStr("crap", 7, encoding.Nop)
+
+		nc.prgROMSize = prgROMSize
+		nc.chrROMSize = chrROMSize
+	}
+
+	nc.nes20 = nes20 == 2 // binary 10
+
+	if trainer {
+		nc.trainerSize = 512
+	} else {
+		nc.trainerSize = 0
+	}
+
+}
+
+func decodePRGROM(d *decode.D) {
+	for !d.End() {
+		peek := d.PeekUintBits(8)
+		peekTyp := opMap[peek].Type
+		peekArgL := ArgLength(peekTyp)
+
+		if d.BitsLeft() < int64(1+peekArgL)*8 {
+			d.FieldRawLen("junk", d.BitsLeft())
+		} else {
+
+			d.FieldStruct("instruction", func(d *decode.D) {
+				op := d.FieldU8("op_code", scalar.UintFn(func(s scalar.Uint) (scalar.Uint, error) {
+					opLookup := opMap[s.Actual]
+					s.Sym = opLookup.Name
+
+					return s, nil
+				}))
+
+				typ := opMap[op].Type
+				argL := ArgLength(typ)
+				switch argL {
+				case 1:
+					d.FieldU8("args", GetArgFormatter(typ))
+				case 2:
+					d.FieldU16("args", GetArgFormatter(typ))
+				}
+			})
+		}
+	}
+}
+
+func decodeCHRROM(d *decode.D) {
+	for !d.End() {
+		d.FieldStruct("tile", func(d *decode.D) {
+			lsbs := d.PeekBytes(8)
+			d.SeekRel(64)
+			msbs := d.PeekBytes(8)
+			d.SeekRel(-64)
+			d.FieldU64("data1")
+			d.FieldU64("data2")
+
+			for i := 0; i < 8; i++ {
+				lsb1 := lsbs[i] & 0x01
+				msb1 := msbs[i] & 0x01
+				pxl1 := lsb1 + (msb1 << 1)
+				lsb2 := (lsbs[i] & 0x02) >> 1
+				msb2 := (msbs[i] & 0x02) >> 1
+				pxl2 := lsb2 + (msb2 << 1)
+				lsb3 := (lsbs[i] & 0x04) >> 2
+				msb3 := (msbs[i] & 0x04) >> 2
+				pxl3 := lsb3 + (msb3 << 1)
+				lsb4 := (lsbs[i] & 0x08) >> 3
+				msb4 := (msbs[i] & 0x08) >> 3
+				pxl4 := lsb4 + (msb4 << 1)
+				lsb5 := (lsbs[i] & 0x10) >> 4
+				msb5 := (msbs[i] & 0x10) >> 4
+				pxl5 := lsb5 + (msb5 << 1)
+				lsb6 := (lsbs[i] & 0x20) >> 5
+				msb6 := (msbs[i] & 0x20) >> 5
+				pxl6 := lsb6 + (msb6 << 1)
+				lsb7 := (lsbs[i] & 0x40) >> 6
+				msb7 := (msbs[i] & 0x40) >> 6
+				pxl7 := lsb7 + (msb7 << 1)
+				lsb8 := (lsbs[i] & 0x80) >> 7
+				msb8 := (msbs[i] & 0x80) >> 7
+				pxl8 := lsb8 + (msb8 << 1)
+				d.FieldValueStr(fmt.Sprintf("pixel_data_%d", i), fmt.Sprintf("%02b %02b %02b %02b %02b %02b %02b %02b",
+					pxl8, pxl7, pxl6, pxl5, pxl4, pxl3, pxl2, pxl1,
+				))
+			}
+		})
+	}
+}
+
+func decodeNES(d *decode.D) any {
+	var nc nesContext
+
+	d.Endian = decode.LittleEndian
+
+	d.FramedFn(16*8, func(d *decode.D) {
+		d.FieldStruct("header", func(d *decode.D) { decodeFileHeader(d, &nc) })
+	})
+
+	if nc.trainerSize > 0 {
+		d.FramedFn(int64(nc.trainerSize)*8, func(d *decode.D) {
+			d.FieldArray("trainer", func(d *decode.D) { decodePRGROM(d) })
+		})
+	}
+
+	d.FramedFn(int64(nc.prgROMSize)*8, func(d *decode.D) {
+		d.FieldArray("prg_rom", func(d *decode.D) { decodePRGROM(d) })
+	})
+
+	d.FramedFn(int64(nc.chrROMSize)*8, func(d *decode.D) {
+		d.FieldArray("chr_rom", func(d *decode.D) { decodeCHRROM(d) })
+	})
+
+	if d.BitsLeft() > 0 {
+		d.FieldRawLen("the_rest", d.BitsLeft())
```

Comment:
Is it know as this? :) or could be "unknown"?

Reply:
Could just as well be unknown. Or maybe I should just throw an error at this point since there isn't supposed to be anything beyond the CHR ROM.

Reply:
I see, in most format decoders if there is some clear "end" signal i've usually just finish successfully and if there is something trailing there will be gap fields added automatically for it.